### PR TITLE
Issue #20: Possible to remove private AS for IBGP peer

### PR DESCRIPTION
**Pull request by lyubov-popova**
Initial commit for BGP commands validation (added validation of remove-private-AS).

Issue #20: Possible to remove private AS for IBGP peer
Change-Id: Ic5fcf894f6c509e888256b59b849f7cc8f9437c6

### CODE REVIEW

File: vtysh/bgp_vty.c [hunk 1]
```diff
@@ -3910,33 +3933,72 @@ cli_neighbor_remove_private_as_cmd_execute(struct vty *vty,
...
                                               &remove_private_as, 1);
-    END_DB_TXN(txn);
+
+    return BGP_SUCCESS;
+}
+
+static int
+bgp_handle_db_transaction(struct ovsdb_idl_txn *txn, int result)
+{
+    if (result == BGP_SUCCESS)
```
Comment (AKochin) [inline vtysh/bgp_vty.c]: How about case (result) here?

File: vtysh/bgp_vty.c [hunk 2]
```diff
@@ -3910,33 +3933,72 @@ cli_neighbor_remove_private_as_cmd_execute(struct vty *vty,
...
+    {
+        ABORT_DB_TXN(txn, "command exists");
+    }
+    else if (result == BGP_ERR_REMOVE_PRIVATE_AS)
+    {
+        ABORT_DB_TXN(txn, "Private AS cannot be removed for IBGP peers");
+    }
+    else
+    {
+        vty_out(vty, "Unknown BGP error code = %d\n", result);
```
Comment (stroykov) [inline vtysh/bgp_vty.c]: transaction is lost here?

**Review by stroykov — CHANGES_REQUESTED** [review]
leave only the changes needed to fix the issue.
restore  other source code

**Comment by AKochin** [comment]
This commit is a part of a big task of validation of BGP commands got from CLI. These changes are start point of it so this code just need to be adjusted not removed.

**Comment by lyubov-popova** [comment]
Hi,
Comments after review were applied. Can you please review it again?
Summary of changes:
1) Validation for remove-private-AS command (check for IBGP peer)
2) remove-private-AS is deleted if remote AS is changed from external to internal
3) Error code constants are introduced instead of repeated error text
4) Check for ovs_bgp_neighbor was added if missed. Using ABORT_DB_TXN if BGP neighbor is not configured:
         ovs_bgp_neighbor = get_bgp_neighbor_with_bgp_router_and_ipaddr(bgp_router_context, peer_str);
         if (ovs_bgp_neighbor) {
              /* write to the DB */
              ovsrec_bgp_neighbor_set_bfd_fallover_enable(ovs_bgp_neighbor, &is_set_bfd, 1);		              
          }
         **else {
             ABORT_DB_TXN(txn, BGP_ERR_NEIGHBOR_IS_NOT_CONFIGURED);
 	}**

File: vtysh/bgp_vty.c [hunk 3]
```diff
@@ -2694,6 +2694,20 @@ bgp_neighbor_peer_group_insert_to_bgp_router(const struct ovsrec_bgp_router *
     free(bgp_neighbor_peer_group_list);
 }
 
+static bool
+is_ibgp_peer(const struct ovsrec_bgp_neighbor *ovs_bgp_neighbor,
```
Comment (stroykov) [inline vtysh/bgp_vty.c:2698]: from the name semantic, the function must take bgp and neighbor input parameters and decide, if that neighbor is of ibgp type.

Comment (lyubov-popova) [reply vtysh/bgp_vty.c:2698]: Do you mean changing parameters in the function or function name?

Tried to change parameters in following way:
```
static bool is_ibgp_peer(const struct ovsrec_bgp_neighbor *ovs_bgp_neighbor,
                                      const struct ovsrec_bgp_router *bgp_router_context)
```

But, as I understand, struct ovsrec_bgp_router does not contain asn, it is stored in `ovs_vrf->key_bgp_routers[j]`
E.g., in vtysh_router_context_bgp_clientcallback function https://github.com/MERAprojects/ops-cli/blob/meraswitch/devel/master/vtysh/vtysh_ovsdb_router_context.c

I can write a comment describing parameters in is_ibgp_peer function.

Comment (stroykov) [reply vtysh/bgp_vty.c:2698]: Leave as it is, but change parameter name asn to own_asn

**Review by stroykov — APPROVED** [review]

**Pull request merged**